feat: LEAP-1469: Implement universal linking modes

Gondragos opened the pull request:
This PR provides the system of universal linking modes. It's the ideological heir of `Relation Mode` with reusing a lot of its own code with a couple of renamings and refactoring. 
As the result we have separated `RelationMode` and `CommentMode`. The second one will provide ability to link comments to regions.

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend



### Describe the reason for change
We need a new mode mostly simmilar to `RelationMode`. That causes refactoring to reuse most of the functionality.



#### What feature flags were used to cover this change?
We avoided to using feature flags here as it might not affect the existing functionality.




### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [x] Not sure (briefly explain the situation below)

It can if there are some problems with code 😅

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [x] integration
- [ ] unit



### Which logical domain(s) does this change affect?
`LinkingModes`, `RelationMode`, `CommentMode`

Gondragos committed:
Add Linking Modes for Annotations

Introduce `CommentMode` and `RelationMode` to handle linking annotations with regions. Also, refactor parts of the code to unify the method of linking actions under `LinkingModes` and replace deprecated `relationMode` checks.

Gondragos committed:
Add integration tests for linking modes to comments

Gondragos committed:
Refactor CommentMode and RelationMode to use volatile state.

Moved the `comment` and `firstRegion` variables to volatile state within the MobX-state-tree models for `CommentMode` and `RelationMode`. This change provides access to payload information of modes in case we need it.

hlomzik commented on web/libs/editor/src/stores/Annotation/LinkingModes.js:
```diff
@@ -0,0 +1,81 @@
+      // @deprecated
+      startRelationMode(obj) {
+        console.warn(
+          "`startRelationMode` is deprecated. Use `startLinkingMode` with `CREATE_RELATION_MODE` as first instead.",
```
```suggestion
          "`startRelationMode` is deprecated. Use `startLinkingMode(CREATE_RELATION_MODE, obj)` instead.",
```

hlomzik commented on web/libs/editor/src/stores/Annotation/LinkingModes.js:
```diff
@@ -0,0 +1,81 @@
+          self.currentLinkingMode.addLinkedRegion?.(region);
+        }
+      },
+      // @deprecated
```
```suggestion
      // @deprecated Use `startLinkingMode(CREATE_RELATION_MODE, obj)` instead
```

hlomzik commented on web/libs/editor/src/stores/Annotation/LinkingModes.js:
```diff
@@ -0,0 +1,81 @@
+        );
+        self.startLinkingMode(RelationMode.key, obj);
+      },
+      // @deprecated
```
```suggestion
      // @deprecated Use `stopLinkingMode` instead
```

hlomzik commented on web/libs/editor/src/stores/Annotation/LinkingModes.js:
```diff
@@ -0,0 +1,81 @@
+  .actions((self) => {
+    return {
+      afterCreate() {
+        self.linkingModes.set(RelationMode.key, RelationMode.model.create({}));
+        self.linkingModes.set(CommentMode.key, CommentMode.model.create({}));
```
Can we declare them statically in `model()`?

Gondragos replied:
I afraid that they could be rewritten as they live in the `model` and all fields there are affected by the data we provide on creating the store.

hlomzik replied:
that's definitely something we should not worry about. on a level of Annotation store we connect all other store parts via model and they are fully controlled by us (in contrary to tag models for example, which are instantiated from config). so better to move this to model for consistency and readability

hlomzik commented on web/libs/editor/tests/integration/data/linked_modes/to_coments.ts:
`to_comments` — one `m` is missing

Commit:
Fix warning message

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Gondragos committed:
Rename misnamed 'to_coments' to 'to_comments'

Gondragos committed:
Change cursor constants for linking mode

Updated all instances of `Constants.RELATION_MODE_CURSOR` to `Constants.LINKING_MODE_CURSOR` across various files. This ensures consistency in naming and improves code readability for linking mode features.

Gondragos commented on web/libs/editor/src/tags/object/RichText/model.js:
```diff
@@ -413,7 +413,7 @@ const Model = types
 
-        if (region.annotation.relationMode) {
+        if (region.annotation.isLinkingMode) {
           region.setHighlight(true);
```
Done. Is it ok now?

Gondragos committed:
Refactor LinkingModes model

Gondragos committed:
Fix linkingModes initing

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10939426819)
> Successfully merged:  create mode 100644 web/libs/editor/tests/integration/e2e/control_tags/classification/textarea.cy.ts

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1469/linking-modes'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10939426819
